### ISSUE TITLE
Open-Vocabulary Detection (OVD) 

### ISSUE THREAD

**Issue by djamahl99** [Luoyadan/Research-Group-Progress#17]
# Frozen VLM features show suitability to open-vocabulary detection
![image](https://user-images.githubusercontent.com/19633934/233239231-900de31e-edee-4c0e-a5ac-f7850303037f.png)

## Current Issues

(Mostly based off of CORA)

![image](https://user-images.githubusercontent.com/19633934/232638319-25ca71fa-5efa-4448-8db7-12926a795ea4.png)
(image from RegionCLIP shows bad performance of CLIP on cropped regions)

### How to adapt CLIP for region-level tasks?

* trivial: crop regions and feed to clip
    * *adopted by MEDet, ViLD, LV-DETR, Object-Centric-OVD*
    * **limitation**: the distribution gap between region crops and full images leads to inferior classification accuracy

![image](https://user-images.githubusercontent.com/19633934/232639402-e14c21f0-8c24-4fa0-8f7f-f273a95654b2.png)
(Image from ViLD showing a RoIAlign being used to acquire regional features)

* crop regions and feed to clip, but augment the text feature and image features
    * *adopted by MEDet*
    * **limitation**: requires extra image-text pairs to prevent overfitting to "base" classes seen during training

![image](https://user-images.githubusercontent.com/19633934/232637518-573e0f9c-ffce-439e-95db-99e31f5140f1.png)
(image from MEDet)

* use RoIAlign to acquire regional features
    * *adopted by RegionCLIP, ViLD, F-VLM*
    *  **limitation**: more efficient, but cannot generalize well to novel classes without finetuning, finetuning is costly when adopting a larger CLIP model     
* use RoIAlign to acquire regional features (from CLIP RN50 layers 1-3) and then pool intermediate feature map using layer 4 and use region prompting to adapt the CLIP image encoder
    * *adopted by CORA*
    * they attribute the generalization capability to the fact that region prompting directly account for the distribution mismatch right after where it occurs (after region pooling), whereas existing methods tweak irrelevant parameters to compensate for the distribution shift
    * **limitation**: RoIAlign cannot be applied to ViT, as not feature grid but sequence

### How to learn generalizable object proposals?
* use RPN or class-agnostic object detectors to mine potential novel class objects
    * *adopted by ViLD, OV-DETR, Object-Centric-OVD, RegionCLIP*
    * RPNs strongly biased towards the base classes on which they are trained
* use handcrafted policiess to rule out or merge low-quality boxes
    * *adopted by MEDet and VL-PLM*
    * **limitation**: performance is still bounded by the frozen RPN
* learn generalizable object localization by conditioning box regression on class name embeddings (Conditional Matching)
    * *adopted by OV-DETR*
    * **limitation**: inefficient due to repetitive per-class inference.
    * during training the number of negative classes sampled in each iteration is limited due to the memory constrain, which hinders convergence

![image](https://user-images.githubusercontent.com/19633934/232976340-ce25a253-7df8-4e12-a7a4-460310fa0bec.png)

* pre-match dynamic anchor boxes with the input categories (Anchor Pre-Matching)
    * *adopted by CORA*
    * adaptively assigned anchor boxes for different classes adaptively according to the image content, ensuring a constant number of query decoupled from the category size. All th
    * **limitation**: 
 
## Limitations of CORA
- only applies to RN50x4 (as otherwise unable to get spatial features maps)
- how could we adapt a ViT to produce feature maps for this task?

## Interesting Techniques
| Paper | Technique | Description |
|------- | ----------- | ------------ |
| CORA | Class dropout | to aid the generalization capability, random categories are randomly dropped. In each iteration, the base classes are split into two complementary sets of categories and train the detector on the same image in the two separate instances. This enforces the model to condition its prediction on the query categories. Since the ground truth box in one group does not apply to the other, the model needs to be aware of the input categories to propose appropriate boxes. |
| CORA | Clip-Aligned labelling | Directly training the localiser on COCO does not converge well. A ground-thruth box can only be matched if a given query box with the same label exists. This means that, GT boxes might be entirely ignored. This is probably due to having innaccurate anchor boxes. CORA relabels the boxes in the training dataset with their region classifier. |
| MEDet | noise removal | |
| ViLD | Including muiltiple crops | Ensembling CLIP predictions (in the trivial cropping design) can produce better classification performance. ![image](https://user-images.githubusercontent.com/19633934/233231500-dea49759-4640-429b-b531-bb2996bc729d.png)
|

# Summary

## Variations

1. train localisation on base classes with NO information about novel classes (no novel text representation)
2. train localisation on base classes with SOME information about novel classes (e.g. text representation)

## Metrics
1. LVIS APr: mAP on LVIS rare classes. The 1203 LVIS classes are separated into 866 frequent/base and 337 novel/rare categories. Some of the rare categories only have very few examples
2. COCO: the 80 classes are split into 48 base categories and 17 novel. 

#### Seen
- toilet
- bicycle
- apple
- train
- laptop
- carrot
- motorcycle
- oven
- chair
- mouse
- boat
- kite
- sheep
- horse
- sandwich
- clock
- tv
- backpack
- toaster
- bowl
- microwave
- bench
- book
- orange
- bird
- pizza
- fork
- frisbee
- bear
- vase
- toothbrush
- spoon
- giraffe
- handbag
- broccoli
- refrigerator
- remote
- surfboard
- car
- bed
- banana
- donut
- skis
- person
- truck
- bottle
- suitcase
- zebra
- background

#### Unseen
- umbrella
- cow
- cup
- bus
- keyboard
- skateboard
- dog
- couch
- tie
- snowboard
- sink
- elephant
- cake
- scissors
- airplane
- cat
- knife

| Paper      | Conference | Implementation                                                                                                                                                          | Novel Information                  | Limitations                          | Opportunities          | LVIS APr    | LVIS AP Novel | COCO Novel AP50 |
|------------|------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|------------------------------------|--------------------------------------|------------------------|-------------|---------------|-----------------|
| Crops      | -          | CLIP on cropped regions                                                                                                                                                 | -                                  | -                                    | -                      | 18.9        |               | 26.3            |
| CORA       | CVPR2023   | anchor pre-matching & region prompting                                                                                                                                  | CLIP + Base labels (CORA+ -> text) | - not applicable to ViT?             | - Stronger backbone    | 28.1        | -             | 43.1            |
| OADP       | CVPR2023   | object-aware knowledge extraction () and distillation pyramid (global and block -> object)                                                                              |                                    |                                      |                        | 21.9        | -             | 35.6            |
| DetCLIPv2  | CVPR2023   | open vocab object detection with word-region alignment using concept enrichment                                                                                         | GLIP + Caption                     | -trained from scratch for grounding  |                        | 43.1        | -             |                 |
| GLIP-L     | CVPR 2022  | phrase grounding (alignment between image regions and words), masked language modeling                                                                                  | GLIP + Caption                     |                                      |                        | 28.2        | -             |                 |
| ViLD       | ICLR 2022  | crop regions and feed to CLIP for image embeddings to distill into a two stage detector with dual heads (text and image). Image head is distilled with image embedding. |                                    | - weak performance on base classes - | - crop ensemble        | 16.3        | 16.8          | 27.6            |
| DetPro     | CVPR 2022  | ViLD + prompt representation learning: 1) background interpretation from prompting 2) context grading to maximise positive proposals                                    |                                    |                                      |                        | 20.8        |               |                 |
| F-VLM      | ICLR 2023  | pre-trained VLM model (fixed) and finetune a detector (Mask RCNN) ontop                                                                                                 |                                    | - lots of hyperparams                |                        | 26.3 (RNx4) |               | 31.4            |
| OV-DETR    | ECCV 2022  | conditional DETR queries on image/text embedding from CLIP (can give text or image)                                                                                     |                                    |                                      |                        | 17.4        | -             | 29.4            |
| MEDet      |            | concepts augmentation, noise removal, fragment mergence (merge fragments with high IoU)                                                                                 |                                    |                                      | - using noise removal  | -           | 22.4          | 32.6            |
| RegionCLIP | CVPR 2022  |                                                                                                                                                                         | CLIP + Caption                     |                                      |                        | 22.0        | 17.1          | 39.3            |
| Detic      | ECCV 2022  |                                                                                                                                                                         |                                    |                                      |                        | -           | 21.0          | 27.8            |
| OVR-CNN    | CVPR 2021  | Established the setting,                                                                                                                                                | CLIP + Caption                     |                                      |                        | -           | -             | 22.8            |

## Risks
- Lots of papers already used CLIP, and whilst could make some decent improvements, probably new GLIP / CLIP will push the SOTA 
- CLIP is aligned strongly with text, but due to this is is also weak and can be strongly enouraged by rendered text [Multimodal Neurons Interactive](https://distill.pub/2021/multimodal-neurons/) and [YouTube Demo, (interactive demo is now frozen)](https://www.youtube.com/watch?v=Z_kWZpgEZ7w) 
![image](https://user-images.githubusercontent.com/19633934/233229997-5abdc9eb-f141-44f7-9476-8f8265b47cad.png)
and I imagine something stronger is going to come out soon / have GLIPv2 now etc. as stronger foundation for open-vocabulary detection
- highly competitive :(
